### PR TITLE
route model access through the thunder module

### PR DESCRIPTION
**Pull request by t-vi**
...then we can override parameters etc. on the thunder module.

### CODE REVIEW

**Comment by nikitaved** [comment]
I was just about to file an issue about that :)

**Comment by t-vi** [comment]
I'll need to fix that some things (such as `__globals__`) still need to go directly onto the original modules.

File: thunder/__init__.py [hunk 1]
```diff
@@ -203,8 +203,18 @@ def __init__(self, model, compiled_model_call):
         super().__init__()
         self._model = model
 
+        del self.training
```
Comment (carmocca) [inline thunder/__init__.py:206]: Why is this here? A comment would be helpful

File: thunder/__init__.py [hunk 2]
```diff
@@ -203,8 +203,18 @@ def __init__(self, model, compiled_model_call):
...
         self._forward_fn = compiled_model_call
 
+    def get_buffer(self, name):
+        return self._model.get_buffer(name)
+
+    def get_parameter(self, name):
+        return self._model.get_parameter(name)
+
+    def get_submodule(self, name):
+        return self._model.get_submodule(name)
```
Comment (carmocca) [inline thunder/__init__.py:216]: We might want to add logic to `__getattr__` so that if the method comes from `nn.Module` and is not overridden, we automatically route to `self._model`

Avoiding the need for these manual additions

File: thunder/__init__.py (hunk 1 above)
Comment (crcrpar) [inline thunder/__init__.py:206]: at glance this reads a tad aggressive. why and how would this be needed?

File: thunder/__init__.py [hunk 3]
```diff
@@ -448,43 +471,44 @@ def get_computation_and_inputs(*args, **kwargs):
...
+                    cs.last_interpreted_instructions = None
+                    cs.last_interpreter_log = None
+                    cs.last_prologue_traces = pro_traces
+                    cs.last_prologue = pro
+                    cs.last_prologue_transformation_start = 0
+                    cs.last_prologue_transformation_stop = 0
+                    cs.last_computation_transformation_start = 0
+                    cs.last_computation_transformation_stop = 0
+
+                    return cache_entry, inps, pro_to_epi
```
Comment (crcrpar) [inline thunder/__init__.py:511]: the diff of these lines is just one indent due to the addition of `compile_data_and_stats` contextmanager?

**Review by crcrpar — COMMENTED** [review]
not a feedback, just a few questions

File: thunder/core/prims.py [hunk 4]
```diff
@@ -689,6 +691,43 @@ def unpack_function_obj_printer(
...
+    )
+    utils.check(
+        len(kwarg_printables) == 0,
+        lambda: f"Expected no kwargs for unpack_thunder_module but got {kwarg_printables}",
+        exception_type=AssertionError,
+    )
+
+    result_str = "_" if bsym.output is None else f"{codeutils.prettyprint(out_printables, with_type=True)}"
+    arg_str = codeutils.prettyprint(arg_printables[0])
+    s = f"{result_str} = thunder._get_thunder_module({arg_str})"
```
Comment (carmocca) [inline thunder/core/prims.py:718]: If this is meant to appear in traces, it's not nice that it's a protected function

File: thunder/__init__.py [hunk 5]
```diff
@@ -692,6 +716,7 @@ def fn_(*args, **kwargs) -> Any:
 
     if isinstance(fn, pytorch.nn.Module):
         fn_ = ThunderModule(fn, fn_)
+        cd._thunder_module_map[id(fn)] = fn_
```
Comment (carmocca) [inline thunder/__init__.py:719]: What about the case that you jit a function that takes the module? We won't have this reference so easily

File: thunder/__init__.py (hunk 1 above)
Comment (t-vi) [reply thunder/__init__.py:206]: I'll add a comment. This is to automatically take the `self._model.training` instead of `self.training`.

File: thunder/__init__.py (hunk 3 above)
Comment (t-vi) [reply thunder/__init__.py:511]: Yes.

File: thunder/__init__.py (hunk 1 above)
Comment (t-vi) [reply thunder/__init__.py:206]: Indeed!

File: thunder/__init__.py (hunk 2 above)
Comment (t-vi) [reply thunder/__init__.py:216]: The idea here is that the next PR will change those to include overrides.

File: thunder/__init__.py (hunk 5 above)
Comment (t-vi) [reply thunder/__init__.py:719]: Yeah, but it needs extension to be of full use, but it is not a regression as is. I added this as a map because it enables the case where we do have modules. We still need to add the mechanism to populate this map.

File: thunder/core/prims.py (hunk 4 above)
Comment (t-vi) [reply thunder/core/prims.py:718]: I'm not keen on having it as a "public api" directly under thunder. Maybe we should just move ThudnerModule to core/module.py and do it there?

Comment (t-vi) [reply thunder/core/prims.py]: I moved things to thunder.core.module now.

**Comment by t-vi** [comment]
Thank you @carmocca @crcrpar
